doc/built-by.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "BUILT-BY 1"
.TH BUILT-BY 1 "2022-03-07" "annobin-1" "RPM Development Tools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
built\-by \- Who Built Me
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
built-by
  [\fB\-\-help\fR]
  [\fB\-\-version\fR]
  [\fB\-\-verbose\fR]
  [\fB\-\-quiet\fR]
  [\fB\-\-silent\fR]
  [\fB\-\-ignore\fR]
  [\fB\-\-readelf=\fR\fIpath\fR]
  [\fB\-\-tmpdir=\fR\fIdir\fR]
  [\fB\-\-tool=\fR\fIname\fR]
  [\fB\-\-nottool=\fR\fIname\fR]
  [\fB\-\-before=\fR\fIdate\fR]
  [\fB\-\-after=\fR\fIdate\fR]
  [\fB\-\-minver=\fR\fIversion\fR]
  [\fB\-\-maxver=\fR\fIversion\fR]
  [\fB\-\-\fR]
  \fIfile\fR...
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fIbuilt-by\fR script reports the name and version of the tool
used to build the specified file(s).  This script also demonstrates
how information can be extracted from other other locations in the
file, not just the binary annotation notes.
.PP
The script can also be used to filter files, only reporting those
built by a specific tool, or a specific version of a tool, or even by
a version of a tool that was built between a range of dates.
.PP
The options available are:
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD 0
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD
Displays the usage of the script and then exits.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD 0
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD
Displays the version of the script.
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD 0
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD
Enables verbose mode, causing the script to detail each action it
takes.
.IP "\fB\-\-quiet\fR" 4
.IX Item "--quiet"
.PD 0
.IP "\fB\-q\fR" 4
.IX Item "-q"
.PD
Do not include the name of script in the out generated by the script.
.IP "\fB\-\-silent\fR" 4
.IX Item "--silent"
.PD 0
.IP "\fB\-s\fR" 4
.IX Item "-s"
.PD
Produce no output.  Just return an exit status.
.IP "\fB\-\-ignore\fR" 4
.IX Item "--ignore"
Do not report file types that do not contain any builder information.
.IP "\fB\-\-tool=\fR\fIname\fR" 4
.IX Item "--tool=name"
Only report binaries built by \fIname\fR.  The \fIname\fR is only an
ordinary string, not a regular expression.
.IP "\fB\-\-nottool=\fR\fIname\fR" 4
.IX Item "--nottool=name"
Skip any binary build by \fIname\fR.  The \fIname\fR is only an
ordinary string, not a regular expression.
.IP "\fB\-\-before=\fR\fIdate\fR" 4
.IX Item "--before=date"
Only report binaries built by a tool that was created before
\&\fIdate\fR.  \fIdate\fR has the format \fI\s-1YYYYMMDD\s0\fR.
.IP "\fB\-\-after=\fR\fIdate\fR" 4
.IX Item "--after=date"
Only report binaries built by a tool that was created after
\&\fIdate\fR.  When combined with the \fB\-\-before\fR option can be
used to restrict output to files which were built by tools created in
a specific date range.
.IP "\fB\-\-minver=\fR\fIversion\fR" 4
.IX Item "--minver=version"
Only report binaries built by a tool whose version is \fIversion\fR or
higher.  The \fIversion\fR string should be in the form \fIV.V.V\fR,
for example \fI6.2.1\fR.
.IP "\fB\-\-maxver=\fR\fIversion\fR" 4
.IX Item "--maxver=version"
Only report binaries built by a tool whose version is \fIversion\fR or
lower.  Can be combined with the \fB\-\-minver\fR option to restrict
output to those binaries created by tools within a specific version
range.
.IP "\fB\-\-tmpdir=\f(BIdir\fB\fR" 4
.IX Item "--tmpdir=dir"
.PD 0
.IP "\fB\-t=\f(BIdir\fB\fR" 4
.IX Item "-t=dir"
.PD
Directory to use to store temporary files.
.IP "\fB\-\-readelf=\f(BIpath\fB\fR" 4
.IX Item "--readelf=path"
.PD 0
.IP "\fB\-r=\f(BIpath\fB\fR" 4
.IX Item "-r=path"
.PD
Use the specified program to read the notes from the files.
.IP "\fB\-\-\fR" 4
.IX Item "--"
Stop accumulating command line options.  This allows the script to be
run on files whose names starts with a dash.
.SH "OPTIONS"
.IX Header "OPTIONS"
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2018 \- 2022 Red Hat.
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the \s-1GNU\s0 Free Documentation License, Version 1.3
or any later version published by the Free Software Foundation;
with no Invariant Sections, with no Front-Cover Texts, and with no
Back-Cover Texts.  A copy of the license is included in the
section entitled \*(L"\s-1GNU\s0 Free Documentation License\*(R".


scripts/built-by.sh
#!/bin/bash

# Script to check which tools built the specified binaries.
#
# Created by Nick Clifton.  <nickc@redhat.com>
# Copyright (c) 2016-2018 Red Hat.
#
# This is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 3, or (at your
# option) any later version.

# It is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# Usage:
#   built-by [options] file(s)
#
# This script does not handle directories.  This is deliberate.
# It is intended that if recursion is needed then it will be
# invoked from find, like this:
#
#   find . -type f -exec built-by.sh {} \;

version=3.1

help ()
{
  # The following exec goop is so that we don't have to manually
  # redirect every message to stderr in this function.
  exec 4>&1    # save stdout fd to fd #4
  exec 1>&2    # redirect stdout to stderr

  cat <<__EOM__

This is a shell script to extract details of the
tool that was used to create the named files.

Usage: $prog {files|options}

  {options} are:
  -h        --help            Display this information.
  -v        --version         Report the version number of this script.
  -V        --verbose         Report on progress.
  -s        --silent          Produce no output, just an exit status.
  -q        --quiet           Do not include the script name in the output.
  -i        --ignore          Silently ignore files where the builder cannot be found.
  -r=<PATH> --readelf=<PATH>  Path to version of readelf to use to read notes.
  -t=<PATH> --tmpfile=<PATH>  Temporary file to use.
  --                          Stop accumulating options.

The information reported can be made conditional by using the following options:

            --tool=<NAME>     Only report binaries built by <NAME>
            --nottool=<NAME>  Skip binaries built by <NAME>
            --before=<DATE>   Only report binaries built before <DATE>
            --after=<DATE>    Only report binaries built after <DATE>
	    --minver=<VER>    Only report binaries built by version <VER> or higher
	    --maxver=<VER>    Only report binaries built by version <VER> or lower

  <NAME> is just a string, not a regular expression
  <DATE> format is YYYYMMDD.  For example: 20161230
  <VER> is a version string in the form V.V.V  For example: 6.1.2

The --before and --after options can be used together to specify a date
range which should be reported.  Similarly the --minver and --maxver
options can be used together to specify a version range.

__EOM__
  exec 1>&4   # Copy stdout fd back from temporary save fd, #4
}

main ()
{
    init
    
    parse_args ${1+"$@"}

    scan_files

    if [ $failed -ne 0 ];
    then
	exit 1
    else
	exit 0
    fi
}

report ()
{
    if [ $silent -ne 0 ];
    then
	return
    fi

    if [ $quiet -eq 0 ];
    then
	echo -n $prog": "
    fi
    
    echo ${1+"$@"}
}

fail ()
{
    report "Internal error: " ${1+"$@"}
    exit 1
}

verbose ()
{
    if [ $verb -ne 0 ]
    then
	report ${1+"$@"}
    fi
}

# Initialise global variables.
init ()
{
    files[0]="";  
    # num_files is the number of files to be listed minus one.
    # This is because we are indexing the files[] array from zero.
    num_files=0;

    failed=0
    silent=0
    quiet=0
    verb=0
    ignore_unknown=0
    scanner=readelf
    tmpfile=/dev/shm/built.by.delme
    tool=""
    nottool=""
    before=""
    after=""
    minver=""
    maxver=""
}

# Parse our command line
parse_args ()
{
    prog=`basename $0`;

    # Locate any additional command line switches
    # Likewise accumulate non-switches to the files list.
    while [ $# -gt 0 ]
    do
	optname="`echo $1 | sed 's,=.*,,'`"
	optarg="`echo $1 | sed 's,^[^=]*=,,'`"
	case "$optname" in
	    -v | --version)
		report "version: $version"
		exit 0
		;;

	    -h | --help)
		help 
		exit 0
		;;
	    -s | --silent)
		silent=1;
		verb=0;
		;;
	    -q | --quiet)
		quiet=1;
		;;
	    -V | --verbose)
		silent=0;
		verb=1;
		;;
	    -i | --ignore)
		ignore_unknown=1;
		;;
	    -r | --readelf)
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "$optname needs a program name"
		    else
			scanner=$1
		    fi
		else
		    scanner="$optarg"
		fi
		;;
	    -t | --tmpfile)
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "$optname needs a filename argument"
		    else
			tmpfile=$1
		    fi
		else
		    tmpfile="$optarg"
		fi
		;;
	    --tool)
		nottool=""
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "$optname needs an argument"
		    else
			tool=$1
		    fi
		else
		    tool=$optarg
		fi
		;;
	    --nottool)
		tool=""
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "$optname needs an argument"
		    else
			nottool=$1
		    fi
		else
		    nottool=$optarg
		fi
		;;
	    --before)
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "$optname needs an argument"
		    else
			before=$1
		    fi
		else
		    before=$optarg
		fi
		;;
	    --after)
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "$optname needs an argument"
		    else
			after=$1
		    fi
		else
		    after=$optarg
		fi
		;;
	    --minver)
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "$optname needs an argument"
		    else
			minver=$1
		    fi
		else
		    minver=$optarg
		fi
		;;
	    --maxver)
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "$optname needs an argument"
		    else
			maxver=$1
		    fi
		else
		    maxver=$optarg
		fi
		;;
	    --)
		shift
		break;
		;;
	    --*)
		report "unrecognised option: $1"
		help
		exit 1
		;;
	    *)
		files[$num_files]="$1";
		let "num_files++"
		;;
	esac
	shift
    done

    # Accumulate any remaining arguments without processing them.
    while [ $# -gt 0 ]
    do
	files[$num_files]="$1";
	let "num_files++";
	shift
    done

    if [ $num_files -gt 0 ];
    then
	# Remember that we are counting from zero not one.
	let "num_files--"
    else
	report "must specify at least one file to scan"
	exit 1
    fi
}

scan_files ()
{
    local i

    i=0;
    while [ $i -le $num_files ]
    do
	scan_file i
	let "i++"
    done
}

scan_file ()
{
    local file

    # Paranoia checks - the user should never encounter these.
    if test "x$1" = "x" ;
    then
	fail "scan_file called without an argument"
    fi
    if test "x$2" != "x" ;
    then
	fail "scan_file called with too many arguments"
    fi

    # Use quotes when accessing files in order to preserve
    # any spaces that might be in the directory name.
    file="${files[$1]}";

    # Catch names that start with a dash - they might confuse readelf
    if test "x${file:0:1}" = "x-" ;
    then
	file="./$file"
    fi

    if ! [ -a "$file" ]
    then
	if [ $ignore_unknown -eq 0 ]; then
	    report "$file: file not found"
	    failed=1
	fi
	return
    fi
    
    if ! [ -f "$file" ]
    then
	if [ $ignore_unknown -eq 0 ]; then
	    report "$file: not an ordinary file"
	    failed=1
	fi
	return
    fi

    if ! [ -r "$file" ]
    then
	if [ $ignore_unknown -eq 0 ]; then
	    report "$file: not readable"
	    failed=1
	fi
	return
    fi

    file $file | grep --silent -e ELF
    if [ $? != 0 ];
    then
	if [ $ignore_unknown -eq 0 ];
	then
	    report "$file: not an ELF format file"
	    failed=1
	fi
	return
    fi
    
    $scanner --wide --notes $file > $tmpfile 2>&1
    if [ $? != 0 ];
    then
	if [ $ignore_unknown -eq 0 ]; then
	    report "$file: scanner '$scanner' failed - see $tmpfile"
	    failed=1
	fi
	# Leave the tmpfile intact so that it can be examined by the user.
	return
    fi

    local -a builder
    local tool_index ver_index date_index tell=1

    grep --silent -e "\$<tool>" $tmpfile

    if [ $? == 0 ];
    then
	# Convert:
	#   $<tool>gcc 7.0.0 20161212 0x00000000 NT_GNU...
	# or
	#   GA$<tool>gcc 7.0.0 20161212 0x00000000 NT_GNU...
	# into:
	#   builder[0]=gcc
	#   builder[1]=7.0.0
	#   builder[2]=20161212

	tool_index=0
	ver_index=1
	date_index=2

	eval 'builder=($(grep -e tool $tmpfile | cut -d " " -f 3-5 | sort -u))'
	
	verbose "build notes contain: ${builder[*]}"
	
	if [ ${#builder[*]} -gt 3 ];
	then
	    report "$file: contains multiple, different creator notes"
	fi

	if [ ${#builder[*]} -lt 3 ];
	then
	    if [ $ignore_unknown -eq 0 ];
	    then
		report "$file: contains truncated creator notes"
		failed=1
	    fi
	    tell=0
	fi
	builder[0]=`echo ${builder[0]} | cut -d \> -f 2`
    else
	verbose "scan for build notes failed, trying debug information"

	# Try examining the debug information in case -grecord-gcc-switches has been used.

	# FIXME: If we have a new enough version of readelf we could add the
	# --debug-dump=follow-links option to cope with separate debug info files..

	$scanner --wide --debug-dump=info $file 2>&1 | grep -e DW_AT_producer > $tmpfile
	eval 'builder=($(grep -e GNU $tmpfile))'

	if [ ${#builder[*]} -ge 11 ];
	then
	    # FIXME: We should grep for the right strings, rather than using
	    # builtin knowledge of the format of the DW_AT_producer contents

	    verbose "DW_AT_producer contains: ${builder[*]}"

	    tool_index=7
	    ver_index=9
	    date_index=10
	    builder[7]="${builder[7]} ${builder[8]}"
	else
	    verbose "scan for debug information failed, trying .comment section"
	    
	    # Alright - last chance.  Check the .comment section
	    $scanner -p.comment $file > $tmpfile 2>&1
	    grep --silent -e "does not exist" $tmpfile

	    if [ $? != 0 ];
	    then
		eval 'builder=($(grep -e GNU $tmpfile))'

		verbose ".comment contains: ${builder[*]}"

		# FIXME: We are using assumed knowledge of the layout of the builder comment.
		if [ ${#builder[*]} -lt 5 ];
		then
		    if [ $ignore_unknown -eq 0 ];
		    then
			verbose "$file: could not parse .comment section"
			report "$file: creator unknown"
			failed=1
		    fi
		    tell=0
		fi
		tool_index=2
		ver_index=4
		date_index=5	
		builder[2]="${builder[2]} ${builder[3]}"
	    else
		if [ $ignore_unknown -eq 0 ];
		then
		    report "$file: creator unknown"
		    failed=1
		fi
		tell=0
	    fi
	fi
    fi

    if [ $tell -eq 1 ];
    then
	if [ x$tool != x ];
	then
	    if [ "${builder[$tool_index]}" == $tool ];
	    then
		tell=0
	    fi
	fi
	if [ x$nottool != x ];
	then
	    if [ "${builder[$tool_index]}" == $nottool ];
	    then
		tell=0
	    fi
	fi
	if [ x$minver != x ];
	then
	    if [[ ${builder[$ver_index]} < $minver ]];
	    then
		tell=0
	    fi
	fi
	if [ x$maxver != x ];
	then
	    if [[ ${builder[$ver_index]} > $maxver ]];
	    then
		tell=0
	    fi
	fi
	if [ x$before != x ];
	then
	    if [ ${builder[$date_index]} -ge $before ];
	    then
		tell=0
	    fi
	fi
	if [ x$after != x ];
	then
	    if [ ${builder[$date_index]} -le $after ];
	    then
		tell=0
	    fi
	fi	
    fi
    
    if [ $tell -eq 1 ];
    then
	report "$file: created by: ${builder[$tool_index]} v${builder[$ver_index]} ${builder[$date_index]}"
    fi

    rm -f $tmpfile
}

# Invoke main
main ${1+"$@"}
